### ISSUE TITLE
Create celery task for processing emails

### ISSUE THREAD

**Issue by mlissner** [freelawproject/courtlistener#1708]
This issue ties together the work in the parser, the model, and the API to actually *do* something. The good news is that we have lots of examples to follow.

**Comment by mlissner** [comment]
So here's the data flow from the moment the CL API gets hit:

1. The API gets a hit and saves the API request to the DB. (The code already does this)

2. From there, the code should generate a chain of celery tasks to take care of things (see line 80 of process_recap_upload for an example of a chain). These are the tasks that need to be completed that I can think of:
      - Parse the email
      - Save the metadata from the email to the DB 
      - Figure out the link to the PDF
      - Download the PDF
      - Extract the PDF to add it to the DB and save it to solr
      - Trigger any alerts, subscriptions and webhooks

    For most of these tasks, we're going to already have code you can either extend or, if we're lucky, simply use:
    - Saving the metadata can probably mirror or possibly use `process_recap_docket`. 
    - Downloading, extracting, and saving the PDF to the DB and solr can probably use a chain like this (from `do_pacer_fetch`):
    ```
    result = chain(
        fetch_pacer_doc_by_rd.si(rd_pk, fq.pk),
        extract_recap_pdf.si(rd_pk),
        add_items_to_solr.si([rd_pk], "search.RECAPDocument"),
        mark_fq_successful.si(fq.pk),
    ).apply_async()
    ```

3. Omitted from the list of things we already have code for is parsing the email (this shouldn't be hard and there are lots of examples), figuring out the link to the PDF (I assume this is part of parsing?), and triggering alerts, subscriptions, and webhooks. Since we already trigger alerts on downloads, I think the trick will be to add a couple hooks to trigger subscriptions and webhooks too. 

So I think that's it. You're diving into a pretty complex part of the code now, so let me know what questions you have. I'll try to support best I can. If you want to do a PR that just has an outline of the functions you think you'll need, with stubs or comments where each piece will go, I bet that'd help.

**Comment by tewen** [comment]
@mlissner Remaining questions at your leisure: https://threadit.app/thread/ozsnywx77iyxpfp4dbg3/message/r3vgzvq4laj6llj3phc62tg5?utm_medium=referral-link

Key questions are how this should relate to RECAPDocument and PacerFetchQueues?

**Comment by mlissner** [comment]
OK, your two questions from the video are:

1. How much of the existing tasks can you run with; and
2. How does logging into PACER work?

For question 1, I'm not sure how to answer without going through each function line by line. I'm happy to do that in a call if that's helpful, but I'll just say one quick thing here, which is that the fetch queue is a separate mechanism for gathering stuff from PACER and storing it in the DB. So you won't want to do anything that touches the fetch queue, but once you have metadata and a PDF, I think you should be able to re-use the processing pipeline. Put another way, it shouldn't matter if you got the PDF and data from an email or by scraping the website — once you have the data, you should be able to ignore where it came from. If this is all too vague or if you want a tour of the data model, I won't blame you. Grab a spot on my calendar. 

For question two, yes, all of the logging in junk should work smoothly. You should be able to prove it to yourself with something like:

```python
from juricraper.pacer import PacerSession
s = Pacersession(username='xyz, password="abc")
s.login()
```

That should work! From there, you'd want to decide what you're scraping and create an object to do so. For example, if you wanted dockets, you could do:

```
from juriscraper.pacer import DocketReport
report = DocketReport(court_id="cand", session=s)
report.query(various=arguments, here=that, i=forget)
print(report.json())  # I think that's right-ish
```

What you want to do is get PDFs. There should be some good examples of doing so in cl.recap.tasks.py if you grep for pdf.

**Comment by tewen** [comment]
Thanks @mlissner just grabbed something for tomorrow at noon.

**Comment by mlissner** [comment]
A couple follow ups from our meeting today...

1. The way to set/cache credentials is via `cl.lib.pacer_session.get_or_cache_pacer_cookies` with signature:

    ```
    def get_or_cache_pacer_cookies(
        user_pk: Union[str, int],
        username: str,
        password: str,
    ) -> RequestsCookieJar:
    ```
    The `user_pk` field can take a username too, and you can see some good examples of it like (which pulls our account from `django.conf.settings`):

    ```
    cookies = get_or_cache_pacer_cookies(
            "pacer_scraper",
            username=settings.PACER_USERNAME,
            password=settings.PACER_PASSWORD,
    )
    ```
    We were looking at this file today, but for some reason it didn't quite click on the call. 

2. Why is the `pacer_case_id` not getting parsed out of HTML in Juriscraper? 

    This actually makes sense in hindsight. The extension can get the pacer_case_id and sends it as an API parameter. With that done, there's no need to parse it, so we don't. In our case, we don't send the `pacer_case_id` as part of the email API POST, so I guess the easiest thing is to add it as something we parse from the HTML, as `docket.pacer_case_id`.

3. Bonus: Where do URLs get parsed? We have several related functions for this, including `juriscraper.pacer.utils....`:
    - `get_court_id_from_url`: Does what's on the tin, if you need the court ID and only have a URL.
    - `get_pacer_case_id_from_nonce_url`: Gets the `pacer_case_id` from a URL like: https://ecf.almd.uscourts.gov/cgi-bin/DktRpt.pl?56120.
    - `get_pacer_seq_no_from_doc1_url`: Gets the sequence number from a doc1 URL.
    - `get_pacer_doc_id_from_doc1_url`: Gets the document ID from the doc1 URL.

4. What Celery tasks can be reused?
    - To download the PDF, we have a nice task that should work very well: `cl.corpus_importer.tasks.download_pacer_pdf_by_rd`. This is what the PDF fetch function, `fetch_pacer_doc_by_rd` uses under the hood. If you get/create the RD first, you should be able to drop this in.
    - `cl.corpus_importer.tasks.update_rd_metadata` is a good one to call next to merge the downloaded and new data into the DB.
    - We'll get some docket metadata from the email that we will want to incorporate. To do that, we'll want a function like `cl.recap.tasks.process_recap_docket` that uses our new email parser data instead of the docket parser, or we'll want to update `process_recap_docket` to be more general purpose (or maybe break it up, I'm not sure which would work out best). 

        That function also does the alert sending and things like that, so you can see where those hooks would fit in. That function is where I'd start my design, I think, plugging in the bits about downloading PDFs and merging RD data above.
    - Finally, you'll need to extract the PDF and update Solr. You can do that with `extract_recap_pdf` and `add_items_to_solr` as mentioned above. 

I *think* that's going to be the right pieces of the puzzle for you, but this is a pretty special case because we haven't gotten docket and PDF information at the same time before.

**Comment by tewen** [comment]
@mlissner So, after my work in #1713 We have the final piece of this to go, which is the "first look" downloads. Here is what I expect to do with that:


## Steps
1. We have saved a large number of the first look HTML pages in a Sanford Law Firm S3 bucket. I can use several of these to test the link finder to make sure it's resilient. The biggest issue with first look is that it's available only once. These samples are really great to have, but it's still hard to ensure anything works until we see it in prod.
2. I anticipate this will run as a separate task in the chain. 
3. There are other things we could do with the emails if first look is not available. However, I would like to stave off scope creep here, and focus on that part of the process. That said, I think it makes sense for you to have a roadmap for the next things you want to do with that. Here are some of the things we have already seen on the Sanford side:
     1. Attorneys there wanted records of "No Document" emails as well as documents. Maybe the Docket model already accounts for this. But there is demand here.
     1.  There is also a restricted document type (I have not seen a lot of these), and they are obviously not available to a system like this. In this case, the attorney will want to know that a restricted doc came in, and still get the first look with their login.
     

## Questions
1. Do we want the document_url property to live on the DocketEntry object? It seems like the most sensible thing, that way we can run a separate task that goes through the DocketEntries and downloads all available first look documents.
2. Should the parsing for the first look page live in Juriscraper? The parser would be fairly simple, as it only has to find one link, but if you want to maintain that separation, we can do it.

**Comment by mlissner** [comment]
>     1. These samples are really great to have, but it's still hard to ensure anything works until we see it in prod.

Makes sense.

>     2. I anticipate this will run as a separate task in the chain.

Yes, seems right. The one concern I have is that we don't want to send alerts/emails/webhooks until after the PDF is downloaded. The current processing code you have (that I just reviewed) sends out alerts, I believe, but if we don't have the PDFs yet, we're in trouble if we do them as the next step in the chain. 

There are a couple ways to fix this:

1. Just don't use a celery chain and instead put the PDF downloading code as a step in the current task for ingesting the docket. This isn't great, but I suppose it would work.

2. Split out the alerting/webhooking/etc. code into a separate task and use that as the chain. So the chain becomes something like:

    ```
    chain(do_email_processing(), get_pdfs(), send_alerts_and_webhooks()).apply_async()
    ```

    I guess that's trickier, and it would scope creep a refactor of the other tasks into this, which I don't love.

    What do you think?


>     3. There are other things we could do with the emails if first look is not available. However, I would like to stave off scope creep here, and focus on that part of the process. That said, I think it makes sense for you to have a roadmap for the next things you want to do with that. Here are some of the things we have already seen on the Sanford side:
>        
>        1. Attorneys there wanted records of "No Document" emails as well as documents. Maybe the Docket model already accounts for this. But there is demand here.

I think it makes sense to send emails in that case. Otherwise, we just swallow the notification, which wouldn't be acceptable. These are usually called "Text only" entries, and we do have them in the model pretty regularly. Not including them in the process now would be a big omission.

>        2. There is also a restricted document type (I have not seen a lot of these), and they are obviously not available to a system like this. In this case, the attorney will want to know that a restricted doc came in, and still get the first look with their login.

Yes, this sounds right. It'd be important to do something here too. I guess a general rule should be that if we get a notification email, we should send an email.

> ## Questions
> 
>     1. Do we want the document_url property to live on the DocketEntry object? It seems like the most sensible thing, that way we can run a separate task that goes through the DocketEntries and downloads all available first look documents.

So I was looking at `cl.recap.tasks.fetch_pacer_doc_by_rd`, which is a function that goes and gets a PDF on behalf of an API request. I think we could (ab)use this to do what we need. Here's the first part of the task:

    ```
    def fetch_pacer_doc_by_rd(self, rd_pk: int, fq_pk: int) -> Optional[int]:
        """Fetch a PACER PDF by rd_pk
    
        This is very similar to get_pacer_doc_by_rd, except that it manages
        status as it proceeds and it gets the cookie info from redis.
    
        :param rd_pk: The PK of the RECAP Document to get.
        :param fq_pk: The PK of the RECAP Fetch Queue to update.
        :return: The RECAPDocument PK
        """
        rd = RECAPDocument.objects.get(pk=rd_pk)
        fq = PacerFetchQueue.objects.get(pk=fq_pk)
        mark_fq_status(fq, "", PROCESSING_STATUS.IN_PROGRESS)
    ```

    This doesn't use a URL parameter, but instead reconstructs the URL based on data it has in the DB already. I think we could do that too, but we'd need the magic variable in the DB for it to work. To accomplish that, we could add a field to the `PacerFetchQueue` table, and then in the second-to-last line above, we could grab the magic variable and use it? 

     I think that'd make this a pretty smooth system that reuses most of our downloading code.

>     2. Should the parsing for the first look page live in Juriscraper? The parser would be fairly simple, as it only has to find one link, but if you want to maintain that separation, we can do it.

Well, doesn't the parser you just put into Juriscraper pull out the magic variable already? If so, we don't need the link because I think we can reconstruct it. Would that work and keep our lives easier?

**Comment by tewen** [comment]
@mlissner 

For the option of "when" to download the PDF, I am pro the simple option, and we just do it there and then. The big refactor is a lot for this one mostly aesthetic thing.

The very last Juriscraper comment refers to parsing the HTML from the first look page. That HTML is different (very different) than other PACER downloads. You have to pull the embeded URL out of this HTML container and then download that. We have ample examples on the Sanford side, but this is a different parsing effort. Here are the options:

1) Recognize that it's very specific to this use case and do it on the CourtListener side.
2) Add it to Juriscraper b/c that's where scraping lives.

Even as I write that down, #2 seems like the more reasonable option. But I figured I would list it here for posterity's sake.

**Comment by mlissner** [comment]
Yes, the parsing code would go in Juriscraper. I don't yet understand why we need that parsing code, but I agree that if we need scraping/parsing code, it should go in Juriscraper. I just don't know what we're pulling from that page that we don't already have from the email itself. The email provides a link...isn't that pretty much all we need?

**Comment by tewen** [comment]
@mlissner Sorry, an example would've been more helpful here. So, I have attached an example of the page you see when you click the first look link.

At least in the case of Sanford, we take a look at this HTML and grab the link for the iFrame: **/cgi-bin/show_temp.pl?file=25392796-0--45540.pdf&type=application/pdf**

```HTML
<html><head><title>Centralized CM/ECF LIVE - U.S. District Court:txwd-Centralized CM/ECF LIVE - U.S. District Court:txwd</title>
<script type="text/javascript">if (top!=self) {top.location.replace(location.href);}</script><link rel="stylesheet" type="text/css" href="/css/default.css"><script type="text/javascript" src="/lib/core.js"></script></head><body><style type="text/css">body { overflow:auto;}</style>
<iframe src="/cgi-bin/show_temp.pl?file=25392796-0--45540.pdf&type=application/pdf" height="100%" width="100%" frameborder="0" scrolling="no">
<a href="/cgi-bin/show_temp.pl?file=25392796-0--45540.pdf&type=application/pdf">click here to view this file</a>
</iframe>
</body></html><SCRIPT LANGUAGE="JavaScript">
		var IsForm = false;
		var FirstField;
		function SetFocus() {
			if(IsForm) {
				if(FirstField) {
					var ind = FirstField.indexOf('document.',0);
					if(ind == 0)
					{
						eval(FirstField);
					}
					else
					{
						var Code = "document.forms[0]."+FirstField+".focus();";
						eval(Code);
					}
				} else {
					var Cnt = 0;
					while(document.forms[0].elements[Cnt] != null) {
						try {
							if(document.forms[0].elements[Cnt].type != "hidden" &&
									!document.forms[0].elements[Cnt].disabled &&
									!document.forms[0].elements[Cnt].readOnly) {
								document.forms[0].elements[Cnt].focus();
								break;
							}
						}
						catch(e) {}
						Cnt += 1;
					}
				}
			}
			return(true);
		}
		</SCRIPT>
</div></body></html>
```

So, it's not heavy parsing, but we do need to grab that out of the page we receive, and then run the download function on that link.

**Comment by mlissner** [comment]
So, take a look at the code here: `juriscraper.pacer.reports.BaseReport.download_pdf`. The iframe thing is something we've seen before, and I think this code will handle it in the courts that do it (not all do, I think). The only thing I don't know about the code that's there is how to adapt it to handle the magic_link junk. Guessing it'd just be another parameter in `juriscraper.pacer.reports.BaseReport._query_pdf_download`, but I haven't seen the HTTP traffic to know for sure. 

Anyway, if we're lucky, the code in the above two locations will (almost) just work for downloading PDFs.

**Comment by mlissner** [comment]
@albertisfu, I wonder if you think this is done?

**Comment by albertisfu** [comment]
@mlissner I think this will be done once #2156 is completed since there were some changes on webhooks triggering.

**Comment by mlissner** [comment]
Great. I'll close it out, then.
